### PR TITLE
🌱 update failuredomain tests in cluster reconciler

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

This PR refactors and adds some tests to the failure domain tests in cluster reconciler.

Changed the InfrastructureObjects to be `GenericInfrastrucutreCluster` objects so that they are more inline with what happens in a real case. In a real worlds case the failure domains are read and updated in the Cluster from InfrastructureCluster Objects and not InfrastructureMachine objects.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/area testing

### CODE REVIEW

**Comment by killianmuldoon** [comment]
I'm missing some context here - I see you're changing the Machine objects here to be Cluster objects, but I don't understand the reason for the change - could you add a line or two in the PR description?

**Comment by ykakarap** [comment]
> I'm missing some context here - I see you're changing the Machine objects here to be Cluster objects, but I don't understand the reason for the change - could you add a line or two in the PR description?

Updated the PR description. TLDR; This is to be more inline with real world case. While the `Kind` is not relevant to the actual test it made more sense to switch to a kind that is more inline with how the controller actually works.

The main intent of the PR is to add a test case that was brought up in https://github.com/kubernetes-sigs/cluster-api/issues/6732.

File: internal/controllers/cluster/cluster_controller_phases_test.go [hunk 1]
```diff
@@ -438,25 +438,32 @@ func TestClusterReconcilePhases_reconcileFailureDomains(t *testing.T) {
...
 			name:                 "expect failure domain to remain same if infra config have same failure domain",
-			cluster:              cluster,
+			cluster:              cluster.DeepCopy(),
 			infraRef:             generateInfraRef(true),
 			expectFailureDomains: newFailureDomain,
 		},
 		{
-			name:                 "expect failure domain to be updated if infra config have update to failure domain",
-			cluster:              clusterWithNewFailureDomainUpdated,
+			name:                 "expect failure domain to be updated if infra config has updates to failure domain",
```
Comment (enxebre) [inline internal/controllers/cluster/cluster_controller_phases_test.go:459]: I think a more appropriate name for `newFailureDomain` would be `infraFailureDomain`.

Also, isn't `expect failure domain to be updated if infra config has updates to failure domain` and `expect failure domain to be reset if infra config have different failure domain` the same test case?

**Comment by sbueringer** [comment]
lgtm pending resolution of https://github.com/kubernetes-sigs/cluster-api/pull/6928/files#discussion_r923267310

**Pull request merged**